### ISSUE TITLE
ThinkPad P50 blackscreen

### ISSUE THREAD

**Issue by thessarb** [Shuzhengz/TPFanCtrl2#23]
hi!
Got an issue with my P50. If i run tpfc on discrete gpu mode the screen goes black and never recovers. If i run it in hybrid mode the discrete gpu crashes and completely disappears from device manager leaving no trace. If i disable dgpu in devmanager, run tpfc and then try to enable dgpu i get error 31 (windows cannot load drivers for this device). This issue persist with every dual fan version of tpfc i found on github and regardless the OS version. BIOS is the latest version (1.67), and ive tried OS versions from W10 1507 up to the latest W11 insider (22563). Im running 22563 currently. 
Thank you in advance

**Comment by Shuzhengz** [comment]
Hmm that's weird, the program doesn't actually access the GPU
Do you have a Xeon CPU or an i7? This might be an issue with how the CPU communicates with the GPU. I don't think this is an issue with the Quadro because the program works with my W540 (i7 X series and K2100M).

But the Nvidia driver (and intel driver) might be a cause so it's probably a good idea to update to the latest version (I also have nvidia's studio driver instead of game ready driver installed).

Also do the single fan versions also have the same issue? (if not then it might be an EC issue) You can find a pretty good list [here](https://thinkwiki.de/TPFanControl)

Thank you!

**Comment by thessarb** [comment]
thank you for your reply
i have the i7-6820HQ with the M1000m. single fan versions of tpfc work fine. the drivers are the latest ones from both intel and nvidia respectively (nvidia-studio drivers).

**Comment by Shuzhengz** [comment]
hm okay, I'll try modding one of the single fan versions into dual fan then

**Comment by thessarb** [comment]
Thank you

**Comment by Remalisa** [comment]
> hm okay, I'll try modding one of the single fan versions into dual fan then

Hello! I have the same problem with black screen on P50. My P50 is Xeon/M2000M. 
I took a lot of time to solve the problem and find a program to control two fans. I tried to diagnose the fans for problems through the official Lenovo Diagnostic Program. When im trying to check the fans I also got a black screen, I hope this information helps to solve the problem.

I really hope for you!

**Comment by Shuzhengz** [comment]
Thank you!

**Comment by Shuzhengz** [comment]
[tpfc.zip](https://github.com/Shuzhengz/TPFanCtrl2/files/8310258/tpfc.zip)
try see if this one works
code on #25

**Comment by Remalisa** [comment]
Hello, I will try it at Monday.

Thanks a lot in advance!!

**Comment by Shuzhengz** [comment]
yeah np!

**Comment by thessarb** [comment]
hi!
just tried it and it still blackscreens. this time windows doesnt seem to freeze on dgpu only mode tho. here are ythe logs captured by the program: 
[TPFanControl.log](https://github.com/Shuzhengz/TPFanCtrl2/files/8310705/TPFanControl.log)
[TPFanControl_csv.txt](https://github.com/Shuzhengz/TPFanCtrl2/files/8310706/TPFanControl_csv.txt)

**Comment by Shuzhengz** [comment]
hmm from the logs it seems like the program crashed while trying to access the fan EC
i think it might be a problem with the the second fan address

**Comment by thessarb** [comment]
thanks for the reply!
is thre anything i could do about it or something oyu would suggest? (fan can be controlled under linux)

**Comment by Shuzhengz** [comment]
yea np!
is it linux mode for the laptop or running GNU/Linux? if it's the ladder do you have the source code for it? the two addresses might be similar
thank you!

**Comment by thessarb** [comment]
it is running GNU/Linux (PopOS 20.10) in tripple boot (also got MacOS Monterey beta). [Here is what i use to control the fans in linux](https://github.com/scientifichackers/thinkfan-control-gui)

**Comment by Shuzhengz** [comment]
oooh yeah i almost forgot that linux doesn't use address to control the EC. I wonder if it's a kernel level thing

**Comment by Remalisa** [comment]
@thessarb Hello! Im sorry that im very noob at programing and understanding of all of this stuff. 
I correctly understand that the problem is not solved at the moment?

**Comment by Shuzhengz** [comment]
yeah, I'm currently working on resolving issue #20 with PR #31, maybe you can try that build out and see if it works

**Comment by thessarb** [comment]
@Remalisa np i am no different
@Shuzhengz thx ill try it out later

**Comment by Shuzhengz** [comment]
heres the link to a wip build
[tpfc.zip](https://github.com/Shuzhengz/TPFanCtrl2/files/8337293/tpfc.zip)

**Comment by thessarb** [comment]
hi!
just tried the wip build and tpfc launches finally, but it lacks cpu temps and cant control the fan. also it shows the fans as if they were stuck at 5120RPM. here is the log:

[TPFanControl.log](https://github.com/Shuzhengz/TPFanCtrl2/files/8445169/TPFanControl.log)

**Comment by Remalisa** [comment]
@thessarb Hi! Can you please explain to me how did you start the program? I was trying, but I failed ;(

**Comment by thessarb** [comment]
@Remalisa i didnt do anything different, just extracted the zip and ran fancontrol.exe (W11 Insider Dev Build 22593)

**Comment by Remalisa** [comment]
Hello! 
I have finally opened the program. I also observe that the temperatures and fan speeds do not change in numbers. I also noticed that when the manual mode is selected from 0 to 7, the fan speed simply does not change. Moreover when you select manual mode 7, it automatically changes to program mode .ini

**Comment by Remalisa** [comment]
@Shuzhengz trying to install rw apears the same problem as in this link: https://www.bleepingcomputer.com/forums/t/709973/rw-everything-not-working-after-windows-build-update-to-the-last-version-1903/

**Comment by Shuzhengz** [comment]
hmm are you on windows 10 or 11?
you can try using another program, any that reads the EC should work, I just haven't used those other than rweverything on windows

**Comment by Remalisa** [comment]
I'm using windows 10 21H2.
Ok, I will try something else, thanks!

**Comment by thessarb** [comment]
hi @Shuzhengz. had too many assignments at uni and didnt have time to track the issue. what program would you recommend to read the EC? i cant seem to find any, or i simply dont know how to search ig

**Comment by Shuzhengz** [comment]
hi sorry, I was very busy with school for the past few weeks and didn't have time to get back to this
rweverything might not work if you have old drivers, so probably a good idea to update all the drivers
I don't know any other equivalent programs, but you could try reading the msr registry using the [rdmsr command](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/rdmsr--read-msr-)
you can also use [msr-tools](https://github.com/intel/msr-tools) provided by Intel, but I'm not sure if that works with Windows (you also have to compile it as there's no binary release)

**Comment by thessarb** [comment]
np, same here, in the middle of an annoying exam season XD. will try it as soon as im done with the exams and post the results here. thanks

**Comment by Lego737** [comment]
Hello Shuzhengz

Is there an update for the P50 with TPFanControl? I've been doing tons of research but can't for the life of me figure it out.

**Comment by Shuzhengz** [comment]
@Lego737 Hi!
sorry I just got back to working on this project, it seems like the version from PR #31 is working better than the intended version, so I'm going to say try that one out, [here's the link to that wip build binary](https://github.com/Shuzhengz/TPFanCtrl2/files/8337293/tpfc.zip)
It seems like it's still reading from the wrong address though, so it would be great if you could try downloading RW everything, then open EC Controller and search for your sensors, since I don't have access to a P50 physically
Thank you

**Comment by Lego737** [comment]
Sure, I will do that then report here!

**Comment by thessarb** [comment]
hi @Shuzhengz! i downloaded rw everything, opened embedded controller but i dont understand the gibberish it puts out. could you give me some hint on how to interpret it? here is an image of what i see:
![image](https://user-images.githubusercontent.com/73614157/175785989-6f050c17-dcbb-4aef-b0b2-d8649640f77a.png)

**Comment by Shuzhengz** [comment]
yup, so currently the release version has the address 0x2F as the data port controller address, which is in hex, and would be 47 in decimal. The number `2F` means that it is on the `20` row and the `0F` column, which would be this one
![image](https://user-images.githubusercontent.com/43869232/176076432-9b1e09db-1b4b-41b6-abff-d82c1225e992.png)

**Comment by Shuzhengz** [comment]
when using manual mode to set fan speed, the value here at `0x2F` would become whatever the inputted fan speed value is (after converting to hex)

**Comment by Shuzhengz** [comment]
try manually change the `0x2F` address to value `46`, if you can hear the speed of the fan ramp up then it's the right address

**Comment by Shuzhengz** [comment]
the temperature readout is currently at address `0x78`, and should be changing in real time
you can find out which address yours is by using another software to monitor your cpu temp (package temp), then convert it to hex, and find which address has the value close o it

**Comment by thessarb** [comment]
So i looked at the said values. 0x78 doesnt really chase the temp reported by the cpu itself, but it matches the cpu temp reported by the EC 1:1. Also 0x79 seems to chase the dGPU temps with 1s delay. Setting 0x2F to 46 ramps up only the left fan.

**Comment by thessarb** [comment]
@Shuzhengz i think i may have found the issue. 0x31 seems to be the "fan selector". When the laptop starts up it hold the value 41 and with that i can control the left fan. With the value 40 i can control the right fan. Dgpu doesnt seem to disappear. However the dual fan tpfc seems to use values 00 and 01 in that address and sure enough, the moment i put either of those values the dgpu goes offline till the next reboot. Also if you leave it on 40 long enough it automarically goes to 00 which causes the dgpu to go offline. However switching between values for slow me seems to have enough time so it doesnt go to 00. Hope this helps.

**Comment by Shuzhengz** [comment]
thank you! I will work on this in a little bit

**Comment by Shuzhengz** [comment]
@thessarb can you try this build? I corrected the ec addresses, however I'm still working on getting the rempeature reading. Currently I'm trying to use [open hardware monitor](https://github.com/openhardwaremonitor/openhardwaremonitor)'s dll in CLI mode to read the package temperature, but there are still a lot of bugs to iron out, so it would be great if you can verify that the fan addresses are working on your laptop, and open hardware monitor also works (you can download the exe [here](https://openhardwaremonitor.org/))

**Comment by Shuzhengz** [comment]
here's the built program: [tpfc-p50b.zip](https://github.com/Shuzhengz/TPFanCtrl2/files/9078094/tpfc-p50b.zip)

**Comment by Lego737** [comment]
Unfourtnately, it sadly didn't work for me. I tried both builds on dGPU mode and the screen still went black.

**Comment by Shuzhengz** [comment]
hmm does it go black immediately or after a few seconds?

**Comment by thessarb** [comment]
@Shuzhengz thank you, ill try it as soon as i get home

**Comment by thessarb** [comment]
@Lego737 boot in hybrid mode, disable dgpu from dev manager so you dont BSOD in a few minutes, open rweverything and choose Embedded Controller, run tpfc. Look at the value at the row 30 and column 01. The value shpuld be 41 at the start, check the value that is input by tpfc.

**Comment by Lego737** [comment]
Yes, it goes black instantly.

**Comment by thessarb** [comment]
@Shuzhengz hi again! im happy to report that tpfc doesnt cause black screen for me anymore. also open hardware monitor seems to work and the values match both hwinfo64 and hwmon

**Comment by Lego737** [comment]
@thessarb, how did you fix the issue in dGPU mode?

**Comment by thessarb** [comment]
I did absolutely nothing. Just tried it in hybrid, dgpu didnt disappear. So i tried it in discrete mode and didnt get a black screen. The temp readout is kinda funky (the degree symbol is weird) but other than that it seems to be working fine for me

**Comment by Lego737** [comment]
I tried his latest build in dGPU mode and my screen still went black.

**Comment by thessarb** [comment]
Try the one Tom uploaded here, thats the one im using

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Shuzhengz** [comment]
@Lego737 (this one)[https://github.com/Shuzhengz/TPFanCtrl2/files/9078094/tpfc-p50b.zip]

**Comment by Shuzhengz** [comment]
@thessarb okay so using C# is giving me a ton of problems, I think I'll try another method by reading the msr

can you try and see from rweverything's cpu msr page if address `0x019C` (the temperature of each core) and address `0x1B1` (package temperature) reports the cpu temperature?

the values would look something like 0000000088260a82, you can ignore everything else and look at the "26" part (decimal digit 16-23 in the more detailed panel), the value should be half of the cpu temperature from what i can tell

thank you!

**Comment by thessarb** [comment]
I will look at it tomorrow after reinstalling windows. Today macos just randomly decided to wipe my windows install clean off the ssd.

**Comment by thessarb** [comment]
hi! sorry for the delay. i opened the msr page but i cant seem to find the said addresses. the list is pretty short and the values dont seem to change at all. a considerable amount of the values are just strings of 0-s.

**Comment by Shuzhengz** [comment]
no worries
the MSR register is actually in the user value section (theres a button on the top right of the window next to the save button to add the addresses), and there should be a value for each of the CPU cores. For 0x019C they should be slightly different, but for 0x1B1 they should all be the same
thank you!

**Comment by Shuzhengz** [comment]
![image](https://user-images.githubusercontent.com/43869232/181603341-6e65242c-ba76-4335-964a-773f664d1b36.png)

**Comment by thessarb** [comment]
![image](https://user-images.githubusercontent.com/73614157/181603456-dc9fbb1e-bb06-4439-8495-264f936481c9.png)
this is all i get there

**Comment by Shuzhengz** [comment]
yeah so you have to add the values in the pop up window
![image](https://user-images.githubusercontent.com/43869232/181603671-aea47203-cf34-4d77-8c89-bb9a44f8ac80.png)

**Comment by Shuzhengz** [comment]
just type in the addresses, click add, and it should show up

**Comment by thessarb** [comment]
just looked at it and i cant relate the values to the core temps in a meaningful way. when the cpu was at 36 degrees, the value there was 42, when the temp was at 60 the value was 3A and when it hit the thermal throttling point it went crazy cycling thru all possible numbers and letter up to F

**Comment by Shuzhengz** [comment]
yeah in Intel's manual there is supposed to be a bitfield modifier for the values to convert it into hex (I believe it's 15:0), I can add that in the code, but I have no idea what it means or actually does.

**Comment by Shuzhengz** [comment]
page 20 of [this](https://www.intel.com/content/dam/www/public/us/en/documents/design-guides/4th-gen-core-family-desktop-tmsdg.pdf)

**Comment by Lego737** [comment]
@Shuzhengz 

I love that you were able to fix this app as I can now use my P50 without it overheating a ton! So, thank you for that!

However, I've come accross a problem that I'm currently looking into. When I launch a game, that being Fortnite, and have the app running in the background, Fortnite ends up crashing after 10 or so minutes. I did, at one point, reinstall the dGPU driver which seemed to fix it but in the end, it still crashed. Do you think that this is the fan app on my machine or could it be something else?! I hope to hear from you soon :)

Lego737

**Comment by Shuzhengz** [comment]
hi @Lego737 
thank you! can you try using a CPU stress test tool like cinebench CPU or blender CPU render, with the fan control set to manual 80 and see if it crashes? this will narrow the issue down to CPU and fan contol or GPU.

also currently I'm still working on temperature for the p50 because it's not in the EC, so it would probably would be a good idea to avoid using smart mode for now and stick to manual mode to avoid overheating

**Comment by btom1970** [comment]
I have a P-50 Xeon M2000. TpFanControl only show CPU and APS temp. This is normal? The others sensors can't be displayed? GPU and others?
I must use the program in manual mode? What values are recommended between 0 and 7?

**Comment by Shuzhengz** [comment]
> I have a P-50 Xeon M2000. TpFanControl only show CPU and APS temp. This is normal? The others sensors can't be displayed? GPU and others? I must use the program in manual mode? What values are recommended between 0 and 7?

unfortunately yes, temps are broken right now as we can't find the address for that

**Comment by AinurSatt** [comment]
Hi @Shuzhengz 

After all this time, I ended up here with a problem on my ThinkPad P50. I tried many versions, but with each one, the laptop wouldn't respond to the program at all.
all i get is:
[13.11.2025 0:52:32] Windows uptime since boot 6471 sec., SecWinUptime= 0 sec.
[13.11.2025 0:52:35] failed to read FanSpeedLowByte 2 from EC
[13.11.2025 0:52:35] failed to read FanSpeedHighByte 2 from EC
[13.11.2025 0:52:35] failed to read FanSpeedLowByte 1 from EC
[13.11.2025 0:52:35] failed to read FanSpeedHighByte 1 from EC
[13.11.2025 0:52:39] failed to read FanSpeedLowByte 2 from EC
[13.11.2025 0:52:39] failed to read FanSpeedHighByte 2 from EC
[13.11.2025 0:52:39] failed to read FanSpeedLowByte 1 from EC
[13.11.2025 0:52:39] failed to read FanSpeedHighByte 1 from EC
[13.11.2025 0:52:43] failed to read FanSpeedLowByte 2 from EC
[13.11.2025 0:52:43] failed to read FanSpeedHighByte 2 from EC
[13.11.2025 0:52:43] failed to read FanSpeedLowByte 1 from EC
[13.11.2025 0:52:43] failed to read FanSpeedHighByte 1 from EC
[13.11.2025 0:52:43] Warning: can't read Status, read error count = 0
[13.11.2025 0:52:43] We will close to BIOS-Mode after 10 consecutive read errors

and when trying to switch to all temperature sensors, the program crashes.
I only managed to control one fan once, but now I can't find the version that worked.

ThinkPad P50 Intel Core i7-6820HQ NVIDIA M2000M

<img width="2147" height="661" alt="Image" src="https://github.com/user-attachments/assets/10379dc0-e8e2-4aa8-b709-1457521ef8a5" />

<img width="2151" height="656" alt="Image" src="https://github.com/user-attachments/assets/f3f57cb4-85bc-4e07-a382-b3c2c239cad5" />

**Comment by FanDjango** [comment]
@AinurSatt Perhaps test one of the bleeding edge tpfancontrol releases for your P50: [here](https://github.com/FanDjango/TPFanCtrl2/releases)

**Comment by Shuzhengz** [comment]
@AinurSatt are you running in hybrid mode or discrete only mode?

**Comment by AinurSatt** [comment]
Using the latest version unfortunately doesn't change anything.

<img width="2150" height="662" alt="Image" src="https://github.com/user-attachments/assets/ca52295f-c0ff-4af4-9de3-b83fb1fdce73" />

**Comment by AinurSatt** [comment]
I'm running in discrete mode only, now I tried it in hybrid mode, but the situation hasn't changed.

**Comment by Shuzhengz** [comment]
Does 2.1.4b work with hybrid mode?
